File: apps/dashboard/lib/secrets.ts
// Encrypt/decrypt per-tenant API credentials using the same scheme the edge
// layer uses (AES-256-GCM keyed by TENANT_SECRET_KEY_B64). Format:
//   "v1:<iv_b64>:<tag_b64>:<ciphertext_b64>"

import { createCipheriv, createDecipheriv, randomBytes } from 'node:crypto';

function key(): Buffer {
  const raw = process.env.TENANT_SECRET_KEY_B64;
  if (!raw) throw new Error('TENANT_SECRET_KEY_B64 missing');
  const buf = Buffer.from(raw, 'base64');
  if (buf.length !== 32) throw new Error('TENANT_SECRET_KEY_B64 must decode to 32 bytes');
  return buf;
}

const VERSION = 'v1';

export function encryptSecret(plaintext: string): string {
  const iv = randomBytes(12);
  const cipher = createCipheriv('aes-256-gcm', key(), iv);
  const ct = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return [VERSION, iv.toString('base64'), tag.toString('base64'), ct.toString('base64')].join(':');
}

export function decryptSecret(blob: string): string {
  const [v, ivB64, tagB64, ctB64, ...rest] = blob.split(':');
  if (v !== VERSION || !ivB64 || !tagB64 || !ctB64 || rest.length > 0) {
    throw new Error('invalid encrypted blob');
  }
  const iv = Buffer.from(ivB64, 'base64');
  const tag = Buffer.from(tagB64, 'base64');
  const ct = Buffer.from(ctB64, 'base64');
  const decipher = createDecipheriv('aes-256-gcm', key(), iv);
  decipher.setAuthTag(tag);
  const pt = Buffer.concat([decipher.update(ct), decipher.final()]);
  return pt.toString('utf8');
}

File: packages/python-shared/veda_shared/infra/secrets.py
"""AES-256-GCM secret decryption — matches the Node `lib/secrets.ts` format.
Format: ``v1:<iv_b64>:<tag_b64>:<ciphertext_b64>``.
Key from env ``TENANT_SECRET_KEY_B64`` (32 raw bytes, base64-encoded).
"""

from __future__ import annotations

import base64
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _key() -> bytes:
    raw = os.environ.get("TENANT_SECRET_KEY_B64")
    if not raw:
        raise RuntimeError("TENANT_SECRET_KEY_B64 not set")
    k = base64.b64decode(raw)
    if len(k) != 32:
        raise RuntimeError("TENANT_SECRET_KEY_B64 must decode to 32 bytes")
    return k


def decrypt_secret(blob: str) -> str:
    """Decrypt a Node-compatible v1 AES-256-GCM blob.

    Raises ValueError on malformed input or decryption failure.
    """
    parts = blob.split(":")
    if len(parts) != 4 or parts[0] != "v1":
        raise ValueError("invalid encrypted blob format")
    iv = base64.b64decode(parts[1])
    tag = base64.b64decode(parts[2])
    ct = base64.b64decode(parts[3])
    aes = AESGCM(_key())
    pt = aes.decrypt(iv, ct + tag, associated_data=None)
    return pt.decode("utf-8")


def encrypt_secret(plaintext: str) -> str:
    """Encrypt to the same wire format the Node side reads."""
    iv = os.urandom(12)
    aes = AESGCM(_key())
    blob = aes.encrypt(iv, plaintext.encode("utf-8"), associated_data=None)
    # AESGCM concatenates ciphertext|tag (last 16 bytes).
    ct, tag = blob[:-16], blob[-16:]
    return ":".join(
        [
            "v1",
            base64.b64encode(iv).decode("ascii"),
            base64.b64encode(tag).decode("ascii"),
            base64.b64encode(ct).decode("ascii"),
        ]
    )
